### PR TITLE
feat(storage): provide end index for compactor iter to reduce io

### PR DESCRIPTION
**Pull request by Li0k**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

The Compactor iterator uses `Unbounded` as the end_bound, so the block stream may read all the block meta.

The pr filters the block_meta by table_id and key_range  to reduce io

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by Li0k** [comment]
test 1000+ compaction group 

with PR (https://github.com/risingwavelabs/risingwave/pull/17235), cn will commit ssts within single group

before optimize

image

after optimize

image

File: src/storage/src/hummock/compactor/iterator.rs [hunk 1]
```diff
@@ -46,7 +46,7 @@ pub struct SstableStreamIterator {
     /// Iterates over the KV-pairs of the current block.
     block_iter: Option<BlockIterator>,
 
-    seek_block_idx: usize,
+    block_idx: usize,
```
Comment (hzxa21) [inline src/storage/src/hummock/compactor/iterator.rs:49]: Please add some documentation for `SstableStreamIterator` and the `block_metas` fields to explain that it may be reading only a sub-set of blocks.

File: src/storage/src/hummock/compactor/iterator.rs [hunk 2]
```diff
@@ -362,48 +361,22 @@ impl ConcatSstableIterator {
...
-                    .contains(&block_metas[start_index].table_id().table_id)
-                {
-                    break;
-                }
-                start_index += &block_metas[(start_index + 1)..]
-                    .partition_point(|block_meta| block_meta.table_id() == start_block_table_id)
-                    + 1;
-            }
-            if start_index >= end_index {
+            let (start_index, end_index) = Self::filter_block_metas(
```
Comment (hzxa21) [inline src/storage/src/hummock/compactor/iterator.rs:364]: How about:
- Passing the ownership of `block_metas` to `filter_block_metas` instead of passing by ref
- Letting `filter_block_metas` returns a `Vec<BlockMeta>` directly

File: src/storage/src/hummock/compactor/iterator.rs [hunk 3]
```diff
@@ -426,6 +399,91 @@ impl ConcatSstableIterator {
...
+                    break;
+                }
+
+                let old_end_index = end_index;
+                let block_metas_to_search = &block_metas[start_index..=end_index];
+
+                end_index = start_index
+                    + block_metas_to_search
+                        .partition_point(|block_meta| {
+                            block_meta.table_id().table_id() < end_block_table_id
```
Comment (hzxa21) [inline src/storage/src/hummock/compactor/iterator.rs:474]: Will the logic here be simpler with two `while`, one for start_index, one for end_index？We can handle both the key range and table id filtering logic together in the while loop.

File: src/storage/src/hummock/compactor/iterator.rs [hunk 4]
```diff
@@ -426,6 +399,91 @@ impl ConcatSstableIterator {
         }
         Ok(())
     }
+
+    pub fn filter_block_metas(
+        block_metas: &Vec<BlockMeta>,
+        seek_key: Option<FullKey<&[u8]>>,
+        existing_table_ids: &HashSet<u32>,
+        key_range: &KeyRange,
```
Comment (hzxa21) [inline src/storage/src/hummock/compactor/iterator.rs:407]: It seems that we assume `seek_key >= key_range.left` and we ignore `key_range.left` in this method. Would it be cleaner if we pass a correct key range for filtering without seek_key here?

File: src/storage/src/hummock/compactor/iterator.rs (hunk 2 above)
Comment (Li0k) [reply src/storage/src/hummock/compactor/iterator.rs:364]: 1. We can't move the block_meta out of the CacheEntry, cloning the whole block_meta in advance is not necessary (input should be ref)
2. We always need to copy the subset, so it makes sense to return Vec.

File: src/storage/src/hummock/compactor/iterator.rs (hunk 3 above)
Comment (Li0k) [reply src/storage/src/hummock/compactor/iterator.rs]: I'm sorry, I didn't fully understand what you meant.

1.The key range filtering and requiring a while loop, which is a one-time operation
2. The table_id filtering is already done with two while loops.

I've refactored the code to make it more readable.

File: src/storage/src/hummock/compactor/iterator.rs (hunk 4 above)
Comment (Li0k) [reply src/storage/src/hummock/compactor/iterator.rs]: Good catch, refactor it

File: src/storage/src/hummock/compactor/iterator.rs [hunk 5]
```diff
@@ -361,49 +365,28 @@ impl ConcatSstableIterator {
...
+
+            let filter_key_range = match seek_key {
                 Some(seek_key) => {
-                    // start_index points to the greatest block whose smallest_key <= seek_key.
-                    block_metas
-                        .partition_point(|block| {
-                            seek_key.cmp(&FullKey::decode(&block.smallest_key)) != Ordering::Less
-                        })
-                        .saturating_sub(1)
+                    KeyRange::new(seek_key.encode().into(), self.key_range.right.clone())
```
Comment (hzxa21) [inline src/storage/src/hummock/compactor/iterator.rs:371]: What happen if `seek_key > self.key_range.right` or `seek_key < self.key_range.left`? It seems that we didn't handle the corner cases when `seek_key` is out of range here.

Comment (zwang28) [reply src/storage/src/hummock/compactor/iterator.rs:371]: How about covering these cases with unit tests?

Comment (Li0k) [reply src/storage/src/hummock/compactor/iterator.rs:371]: UT has cover these two case

And we have handle the `seek_key` and `key_range` in `fn seek` 

```rust
/// Resets the iterator and seeks to the first position where the stored key >= `key`.
    async fn seek<'a>(&'a mut self, key: FullKey<&'a [u8]>) -> HummockResult<()> {
        let seek_key = if self.key_range.left.is_empty() {
            key
        } else {
            match key.cmp(&FullKey::decode(&self.key_range.left)) {
                Ordering::Less | Ordering::Equal => FullKey::decode(&self.key_range.left),
                Ordering::Greater => key,
            }
        };
        let table_idx = self.sstables.partition_point(|table| {
            // We use the maximum key of an SST for the search. That way, we guarantee that the
            // resulting SST contains either that key or the next-larger KV-pair. Subsequently,
            // we avoid calling `seek_idx()` twice if the determined SST does not contain `key`.

            // Note that we need to use `<` instead of `<=` to ensure that all keys in an SST
            // (including its max. key) produce the same search result.
            let max_sst_key = &table.key_range.as_ref().unwrap().right;
            FullKey::decode(max_sst_key).cmp(&seek_key) == Ordering::Less
        });

        self.seek_idx(table_idx, Some(key)).await
    }
```

Comment (zwang28) [reply src/storage/src/hummock/compactor/iterator.rs:371]: It can be `KeyRange::new(max(seek_key.encode().into(), self.key_range.left if not None), self.key_range.right.clone())`

**Review by zwang28 — APPROVED** [review]
LGTM
